fitbenchmarking/parsing/cutest_parser.py

```diff
@@ -0,0 +1,240 @@
+
+        # set the MASTSIF environment variable so that pycutest
+        # can find the sif files
+        os.environ["MASTSIF"] = self.mastsif_dir.name
+
+        self._n = None
+        self._m = None
+
+        # get just the short filename (minus the .SIF)
+        fp = FittingProblem()
+
+        # Collect x and create new file with blank y
+        fname, fp.data_x, fp.data_y, fp.data_e = self._setup_data()
+
+        self._p = self._import_problem(fname)
+
+        fp.name = self._p.name
+        self._y = fp.data_y
+
+        fp.function = self._function  # self._p.obj
+        fp.equation = None
+        fp.starting_values = self._get_starting_values()
+        fp.start_x = None
+        fp.end_x = None
+
+        # Use the object id as a hash to the function.
+        # In order to prevent the id being reused also store the x_data
+        self._cache = {id(fp.data_x): {'f': self._p.objcons, 'x': fp.data_x}}
+
+        return fp
+
+    def __del__(self):
+        self.mastsif_dir.cleanup()
+
+    def _import_problem(self, file_name):
+        """
+        Import the problem using cutest
+
+        :param file_name: The path to the file
+        :type file_name: str
+        :return: The parsed problem
+        :rtype: pycutest.CUTEstProblem
+        """
+
+        problem_ext = os.path.basename(file_name)
+        problem, _ = os.path.splitext(problem_ext)
+
+        return pycutest.import_problem(problem)
+
+    def _function(self, x, *params):
+        """
+        If these x values have been passed in before, then run the function
+        Otherwise, create a new problem file, parse and cache the function
+
+        :param x: The data to evaluate at
+        :type x: np.array
+        :return: The result of evaluating at the given x
+        :rtype: np.array
+        """
+        os.environ["MASTSIF"] = self.mastsif_dir.name
+
+        try:
+            f = self._cache[id(x)]['f']
+        except KeyError:
+            fname, _, _, _ = self._setup_data(x)
+            p = self._import_problem(fname)
+            f = p.objcons
+            self._cache[id(x)] = {'f': f, 'x': x}
+        _, fx = f(np.asarray(params))
+
+        return fx
+
+    def _get_starting_values(self):
+
+        starting_values = [
+            OrderedDict([
+                ('f{}'.format(i), self._p.x0[i])
+                for i in range(self._n)
+            ])
+        ]
+
+        return starting_values
+
+    def _setup_data(self, x=None):
+        """
+        Parses CUTEst SIF files to extract data points
+
+        :param x: X values to save (used for evaluating at new points)
+        :type x: np.array
+        :returns: data_x, data_y
+        :rtype: lists of floats
+        """
+
+        if self.file.closed:
+            with open(self._filename, 'r') as f:
+                lines = f.readlines()
+        else:
+            lines = self.file.readlines()
+
+        if x is not None and len(x.shape) == 0:
```

Comment:
Can `len(x.shape) == 0` ever be `True`?  Even `len(np.array([]))` is 1.   I'm also a bit confused as to what you are trying to achieve here with the cast + additional list.  It feels like you are trying to convert a numeric type like `float` or `int` to a numpy array, although if that is the case `x.shape` would raise an `AttributeError`.

Reply:
`np.array(5)` would have shape `()` so this is just making sure it's a 1-vector instead

Reply:
Ah, fair enough. I didn't even realise that was a possibility, I thought the argument had to be a sequence.  In which case it would be more pythonic (and stop pylint complaining) if you were to do: `if x is not None and len(x.shape)`

Reply:
This doesn't get flagged by my linter... but I'll change it